File: src/main/java/de/tectoast/emolga/commands/showdown/SearchReplaysCommand.java
package de.tectoast.emolga.commands.showdown;

import de.tectoast.emolga.commands.Command;
import de.tectoast.emolga.commands.CommandCategory;
import de.tectoast.emolga.commands.GuildCommandEvent;
import net.dv8tion.jda.api.entities.TextChannel;
import org.jsolf.JSONArray;
import org.jsolf.JSONObject;
import org.jsolf.JSONTokener;

import java.io.IOException;
import java.net.URL;

public class SearchReplaysCommand extends Command {
    public SearchReplaysCommand() {
        super("searchreplays", "Sucht nach Replays der angegebenen Showdownbenutzernamen", CommandCategory.Showdown);
        setArgumentTemplate(ArgumentManagerTemplate.builder()
                .add("user1", "User 1", "Der Showdown-Username von jemandem", ArgumentManagerTemplate.Text.any())
                .add("user2", "User 2", "Der Showdown-Username eines potenziellen zweiten Users", ArgumentManagerTemplate.Text.any(), true)
                .setExample("!searchreplays TecToast")
                .build());
    }

    @Override
    public void process(GuildCommandEvent e) throws IOException {
        TextChannel tco = e.getChannel();
        String url = "https://replay.pokemonshowdown.com/search.json?user=";
        ArgumentManager args = e.getArguments();
        String user1 = args.getText("user1");
        if (args.has("user2")) url += toSDName(user1) + "&user2=" + toSDName(args.getText("user2"));
        else url += toSDName(user1);
        System.out.println(url);
        JSONArray array = new JSONArray(new JSONTokener(new URL(url).openStream()));
        System.out.println(array.toString(4));
        StringBuilder str = new StringBuilder();
        if (array.length() == 0) {
            if (args.has("user2"))
                e.reply("Es wurde kein Kampf zwischen " + user1 + " und " + args.getText("user2") + " hochgeladen!");
            else
                e.reply("Es wurde kein Kampf von " + user1 + " hochgeladen!");
            return;
        }
        for (int i = 0; i < array.length(); i++) {
            JSONObject o = array.getJSONObject(i);
            str.append(o.getString("p1")).append(" vs ").append(o.getString("p2")).append(": https://replay.pokemonshowdown.com/").append(o.getString("id")).append("\n");
        }
        System.out.println(str);
        e.getChannel().sendMessage(str.toString()).queue();
    }
}

File: src/main/kotlin/de/tectoast/emolga/commands/showdown/SearchReplaysCommand.kt
package de.tectoast.emolga.commands.showdown

import de.tectoast.emolga.commands.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import kotlinx.coroutines.delay
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import org.slf4j.LoggerFactory
import java.io.IOException

class SearchReplaysCommand :
    Command("searchreplays", "Sucht nach Replays der angegebenen Showdownbenutzernamen", CommandCategory.Showdown) {
    init {
        argumentTemplate = ArgumentManagerTemplate.builder()
            .add("user1", "User 1", "Der Showdown-Username von jemandem", ArgumentManagerTemplate.Text.any()).add(
                "user2",
                "User 2",
                "Der Showdown-Username eines potenziellen zweiten Users",
                ArgumentManagerTemplate.Text.any(),
                true
            ).setExample("!searchreplays TecToast").build()
        slash(false)
    }

    @Throws(IOException::class)
    override suspend fun process(e: GuildCommandEvent) {
        val args = e.arguments
        val u1 = args.getText("user1")
        val u2 = args.getNullable<String>("user2")
        e.deferReply()
        val body = getBody(u1, u2)
        logger.info(body)
        val jsonstring = if (body.length < 30) {
            e.hook.sendMessage("Verbindung mit dem Showdown-Server fehlgeschlagen, ich versuche es in 10 Sekunden erneut...")
                .queue()
            delay(10000)
            getBody(u1, u2)
        } else body
        try {
            e.hook.sendMessage(
                JSON.decodeFromString<List<Replay>>(jsonstring).take(15)
                    .joinToString("\n") { "${it.p1} vs ${it.p2}: https://replay.pokemonshowdown.com/${it.id}" }
                    .ifEmpty { "Es wurde kein Kampf ${u2?.let { "zwischen $u1 und $it" } ?: "von $u1"} hochgeladen!" })
                .queue()
        } catch (ex: Exception) {
            e.hook.sendMessage("Es konnte keine Verbindung zum Showdown-Server hergestellt werden!").queue()
        }
    }

    private suspend fun getBody(u1: String, u2: String?) = httpClient.get(
        "https://replay.pokemonshowdown.com/search.json?user=${toUsername(u1)}".notNullAppend(u2?.let {
            "&user2=${toUsername(it)}"
        }).also { logger.info(it) }
    ).bodyAsText()

    companion object {
        private val logger = LoggerFactory.getLogger(SearchReplaysCommand::class.java)
    }

    @Serializable
    data class Replay(val id: String, val p1: String, val p2: String, val uploadtime: Long, val format: String)
}